### ISSUE TITLE
Crash on exit due to temporary file

### ISSUE THREAD

**Issue by xdaimon** [qutebrowser/qutebrowser#2168]
QuteBrowser crashes on :wq when a file is automatically opened after being downloaded (control + x).

How to reproduce:
1. Download some file and hit control-x when prompted for the save directory.
2. Let the download finish and launch.
3. When the launched program is still open, execute the :wq command in qutebrowser.

I'm on windows 10 and here is my version info. I installed qutebrowser from one of the downloadable installers on the 'releases' page. Good luck, and thank you for the great browser!

Version info

qutebrowser v0.8.2
Git commit: v0.8.2-dirty (2016-08-02 18:33:59 +0200)

CPython: 3.4.3
Qt: 5.5.1
PyQt: 5.5.1

sip: 4.17
colorama: 0.3.7
pypeg2: 2.15
jinja2: 2.8
pygments: 2.1.3
yaml: 3.11
cssutils: 1.0.1 $Id$
typing: no
PyQt5.QtWebEngineWidgets: no
pdf.js: 1.4.20 (bundled)
Webkit: 538.1
Harfbuzz: system
SSL: OpenSSL 1.0.1i 6 Aug 2014

Style: QWindowsVistaStyle
Platform: Windows-8-6.2.9200, 64bit
Desktop: None
Frozen: True
Imported from C:\Program Files\qutebrowser\library.zip\qutebrowser
OS Version: 8, 6.2.9200, , Multiprocessor Free

**Comment by The-Compiler** [comment]
I can't reproduce this on Linux and Windows 10, even with v0.8.2.

Can you please try v0.8.4, and also show the relevant part of the debug log here (or press report in the report window)?

**Comment by xdaimon** [comment]
I updated to v0.8.4 and the issue still exists for me.

Here is the log [http://pastebin.com/1aBbXYft](url)

It says ,

> ==== Exception ====
Traceback (most recent call last):
  File "scripts/dev\..\..\scripts\..\qutebrowser\commands\runners.py", line 303, in run_safely
  File "scripts/dev\..\..\scripts\..\qutebrowser\commands\runners.py", line 291, in run
  File "scripts/dev\..\..\scripts\..\qutebrowser\commands\command.py", line 531, in run
  File "scripts/dev\..\..\scripts\..\qutebrowser\app.py", line 734, in save_and_quit
  File "scripts/dev\..\..\scripts\..\qutebrowser\app.py", line 667, in shutdown
  File "scripts/dev\..\..\scripts\..\qutebrowser\app.py", line 715, in _shutdown
  File "scripts/dev\..\..\scripts\..\qutebrowser\browser\webkit\downloads.py", line 1304, in cleanup
  File "C:\Users\florian\proj\qutebrowser\.tox\cxfreeze-windows\lib\tempfile.py", line 689, in cleanup
  File "C:\Users\florian\proj\qutebrowser\.tox\cxfreeze-windows\lib\shutil.py", line 478, in rmtree
  File "C:\Users\florian\proj\qutebrowser\.tox\cxfreeze-windows\lib\shutil.py", line 373, in _rmtree_unsafe
  File "C:\Users\florian\proj\qutebrowser\.tox\cxfreeze-windows\lib\shutil.py", line 371, in _rmtree_unsafe
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\xdaimon\\AppData\\Local\\Temp\\qutebrowser-downloads-qv0brn2s\\tmp4r3wfm7vwrar540.exe'
>

**Comment by Kingdread** [comment]
Qutebrowser deletes the files that you open with Ctrl + X on exit, it does not check if they are still being used. In your case, you're running the downloaded file, and Windows doesn't let you delete files that are still being run. It probably won't be reproducible on Linux, as files are handled differently there.

**Comment by The-Compiler** [comment]
I guess we should just ignore (and maybe log) errors when removing temporary download files - we did what we could at that point :wink:

**Comment by The-Compiler** [comment]
huh?

**Comment by Kingdread** [comment]
Seems like this bug is now fixed, 238848903877e41797979dda59ecac2125f6b695 (included in v0.9) introduced a `try/except`  around the cleanup.

The browser doesn't crash anymore, and instead just prints the error to the console (if you even have the console open on Windows):

```
15:32:06 ERROR: Failed to clean up temporary download directory
Traceback (most recent call last):
  File "I:\Daniel\Downloads\qutebrowser\qutebrowser\browser\downloads.py", line 1124, in cleanup
    self._tmpdir.cleanup()
  File "I:\Daniel\Downloads\qutebrowser\.venv\lib\tempfile.py", line 811, in cleanup
    _shutil.rmtree(self.name)
  File "I:\Daniel\Downloads\qutebrowser\.venv\lib\shutil.py", line 488, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "I:\Daniel\Downloads\qutebrowser\.venv\lib\shutil.py", line 383, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "I:\Daniel\Downloads\qutebrowser\.venv\lib\shutil.py", line 381, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] Der Prozess kann nicht auf die Datei zugreifen, da sie von einem anderen Prozess verwendet wird: 'C:\\[...]\\tmp2arvrvkhnpp.7.3.3.Installer.exe'
```
